di-sukharev/opencommit
# feat: Integrate @commitlint for Enhanced Commit Message Generation and Configuration Support

# Description
This pull request integrates the @commitlint module to enhance commit message generation through the adoption of locally defined rules, and to optimize configuration management. This enhancement not only supports switching between different prompt modules but also provides developers with detailed control over commit message structures through the application of commitlint rules.

https://github.com/di-sukharev/opencommit/issues/60
https://github.com/di-sukharev/opencommit/issues/61

## Adaptive Consistency with Commitlint Rules Through a Two-Step Transformation
One of the prime features of this integration is achieving consistency in commit messages according to locally defined dynamic commitlint rules. The process is refined into two transformation steps:

1. **LLM-Compatible Rules Inference**: Initially, commitlint rules are inferred and translated into static prompts that are compatible with LLM.
2. **OpenAI for Example Commit Message Creation**: This step utilizes OpenAI to generate a sample commit message structured in JSON, using a given commit diff. This message should ideally comply with commitlint rules at this stage. 
3. **Injecting Consistency Prompt for few-shots main prompt**: The exemplary commit message generated earlier is utilized as a consistency prompt to guide the LLM in generating the commit message for subsequent commit diffs in a few-shots technique. This ensures that the messages are consistent and adhere to the locally defined commitlint rules. 

## Summary of Changes
- Instructions were added on how to switch to the @commitlint prompt module using the `OCO_PROMPT_MODULE` environment variable. Details on how integration takes place and how to view or modify the configuration have been included.
- The `@commitlint/types` package has been added to devDependencies to ensure compatibility with the conventional commit format. The `crypto` package has been added to dependencies for cryptographic functions (hash comparison).
- Refactor of the structure of diff message generation for readability and facilitated plugging in of prompt modules.
- Added support for the `OCO_PROMPT_MODULE` configuration option allowing users to switch between '@commitlint' or the default 'conventional-commit' as the prompt module.
- Added a commitlint configuration file, utility functions, and types to the commitlint module. This includes utility functions for reading, writing, and checking the existence of the configuration file, and also for cleaning up the configuration string.

## Testing
1. Open a project which contains a commitlint configuration. (See [install step](https://commitlint.js.org/#/?id=install))
2. Set the `OCO_PROMPT_MODULE` environment variable with `oco config set OCO_PROMPT_MODULE=@commitlint` ([Locally of globally](https://github.com/di-sukharev/opencommit#local-per-repo-configuration)).
3. Verify that commit messages are generated according to the specified commitlint rules.

## Example

<img width="797" alt="image" src="https://github.com/di-sukharev/opencommit/assets/5808108/2bff3257-5ed4-4d3e-8655-23c663c8c08d">


File `.commitlintrc.yml`: (volontarily set weird `type-enum` to demonstrate feature)
```md
extends:
  - "@commitlint/config-conventional"
rules:
  body-leading-blank:
    - 2
    - always
  body-max-line-length:
    - 2
    - always
    - 100
  footer-leading-blank:
    - 2
    - always
  footer-max-line-length:
    - 2
    - always
    - 100
  header-max-length:
    - 2
    - always
    - 100
  subject-case:
    - 2
    - never
    - - sentence-case
      - start-case
      - pascal-case
      - upper-case
  subject-empty:
    - 2
    - never
  subject-full-stop:
    - 2
    - never
    - .
  type-case:
    - 2
    - always
    - lower-case
  type-empty:
    - 2
    - never
  type-enum:
    - 2
    - always
    - - featuring
      - fixing
      - documenting
      - styling
      - refactoring
      - testing
      - chores
      - perf
      - build
      - ci
      - revert
  body-empty:
    - 2
    - never
  scope-empty:
    - 2
    - never
prompt:
  questions:
    type:
      description: Select the type of change that you're committing
      enum:
        featuring:
          description: A new feature
        fixing:
          description: A bug fix
        documenting:
          description: Documentation only changes
        styling:
          description: >-
            Changes that do not affect the meaning of the code (white-space,
            formatting, missing semi-colons, etc)
        refactoring:
          description: A code change that neither fixes a bug nor adds a feature
        perf:
          description: A code change that improves performance
        testing:
          description: Adding missing tests or correcting existing tests
        build:
          description: >-
            Changes that affect the build system or external dependencies
            (example scopes: gulp, broccoli, npm)
        ci:
          description: >-
            Changes to our CI configuration files and scripts (example scopes:
            Travis, Circle, BrowserStack, SauceLabs)
        chores:
          description: Other changes that don't modify src or test files
        revert:
          description: Reverts a previous commit
```

Generated prompts at config stage. File `.opencommit-commitlint`:
```json
{
  "hash": "f7473338e1a458cf3c33801f61943d31beb9532377b42e490a6ea6c4bbeb900e",
  "prompts": [
    "There should always be a blank line at the beginning of the body.",
    "Each line of the body should always have 100 characters or less.",
    "There should always be a blank line at the beginning of the footer.",
    "Each line of the footer should always have 100 characters or less.",
    "The header should always have 100 characters or less.",
    "The subject should never be in one of the following case: \n  - sentence-case\n  - start-case\n  - pascal-case\n  - upper-case.",
    "The subject should never be empty.",
    "The subject should never end with '.'.",
    "The type should always be in lower-case case.",
    "The type should never be empty.",
    "The type should always be one of the following values: \n  - featuring (A new feature)\n  - fixing (A bug fix)\n  - documenting (Documentation only changes)\n  - styling (Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc))\n  - refactoring (A code change that neither fixes a bug nor adds a feature)\n  - testing (Adding missing tests or correcting existing tests)\n  - chores (Other changes that don't modify src or test files)\n  - perf (A code change that improves performance)\n  - build (Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm))\n  - ci (Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs))\n  - revert (Reverts a previous commit).",
    "The body should never be empty.",
    "The scope should never be empty."
  ],
  "consistency": {
    "english": {
      "localLanguage": "english",
      "commitFix": "fixing(server): change port variable to uppercase",
      "commitFeat": "featuring(server): allow server to listen on environment variable specified port",
      "commitDescription": "Change port variable to uppercase (#123)\nAllow server to listen on environment variable specified port (#124)"
    }
  }
}
```

On second run, configuration already exists:
<img width="440" alt="image" src="https://github.com/di-sukharev/opencommit/assets/5808108/bc15b9ff-d800-4cc6-b8c1-3a73409bb300">


Please review and provide feedback
